src/library_manager/lib_manager.c

```diff
@@ -0,0 +1,681 @@
+				++idx, ++module_entry) {
+			if (module_entry->type.lib_code) {
+				ret_tmp =
+				lib_manager_unload_module(lib_id << LIB_MANAGER_LIB_ID_SHIFT |
+							  idx, mod, desc);
+				if (ret_tmp < 0)
+					ret = ret_tmp;
+			}
+		}
+	}
+
+	return ret;
+}
+
+static uint8_t *lib_manager_get_instance_bss_address(uint32_t module_id, uint32_t instance_id,
+						     struct sof_man_module *mod)
+{
+	uint32_t instance_bss_size =
+		 mod->segment[SOF_MAN_SEGMENT_BSS].flags.r.length / mod->instance_max_count;
+	uint32_t inst_offset = instance_bss_size * CONFIG_MM_DRV_PAGE_SIZE * instance_id;
+	uint8_t *va_base =
+		(uint8_t *)(mod->segment[SOF_MAN_SEGMENT_BSS].v_base_addr + inst_offset);
+
+	tr_dbg(&lib_manager_tr,
+	       "lib_manager_get_instance_bss_address() instance_bss_size: 0x%x, pointer: 0x%x",
+	       instance_bss_size, va_base);
+	return va_base;
+}
+
+static int lib_manager_allocate_module_instance(uint32_t module_id, uint32_t instance_id,
+						uint32_t is_pages, struct sof_man_module *mod)
+{
+	int ret;
+	uint32_t bss_size =
+			(mod->segment[SOF_MAN_SEGMENT_BSS].flags.r.length / mod->instance_max_count)
+			 * CONFIG_MM_DRV_PAGE_SIZE;
+	uint8_t *va_base = lib_manager_get_instance_bss_address(module_id, instance_id, mod);
+
+	if ((is_pages * CONFIG_MM_DRV_PAGE_SIZE) > bss_size) {
+		tr_err(&lib_manager_tr,  "is_pages (%d) invalid, required: %d",
+		       is_pages, bss_size / CONFIG_MM_DRV_PAGE_SIZE);
+		return -ENOMEM;
+	}
+	/*
+	 * Map bss memory and clear it.
+	 */
+	ret = sys_mm_drv_map_region(va_base, POINTER_TO_UINT(NULL),
+				    bss_size, SYS_MM_MEM_PERM_RW);
+	if (ret < 0)
+		return -ENOMEM;
+	memset(va_base, 0, bss_size);
+	return 0;
+}
+
+static int lib_manager_free_module_instance(uint32_t module_id, uint32_t instance_id,
+					    struct sof_man_module *mod)
+{
+	int ret;
+	uint32_t bss_size =
+			(mod->segment[SOF_MAN_SEGMENT_BSS].flags.r.length / mod->instance_max_count)
+			 * CONFIG_MM_DRV_PAGE_SIZE;
+	uint8_t *va_base = lib_manager_get_instance_bss_address(module_id, instance_id, mod);
+	/*
+	 * Unmap bss memory and clear it.
+	 */
+	ret = sys_mm_drv_unmap_region(va_base, bss_size);
+	return ret;
+}
+
+uint32_t lib_manager_allocate_module(const struct comp_driver *drv,
+				     struct comp_ipc_config *ipc_config,
+				     void *ipc_specific_config)
+{
+	struct sof_man_fw_desc *desc;
+	struct sof_man_module *mod;
+	struct ipc4_base_module_cfg *base_cfg = ipc_specific_config;
+	int ret = -ENOMEM;
+	uint32_t module_id = IPC4_MOD_ID(ipc_config->id);
+	uint32_t entry_index = LIB_MANAGER_GET_MODULE_INDEX(module_id);
+
+	tr_dbg(&lib_manager_tr, "lib_manager_allocate_module() mod_id: 0x%x", ipc_config->id);
+
+	desc = lib_manager_get_library_module_desc(module_id);
+	mod = (struct sof_man_module *)((char *)desc + SOF_MAN_MODULE_OFFSET(entry_index));
+
+	ret = lib_manager_load_module(module_id, mod, desc);
+	if (ret < 0)
+		return 0;
+
+	ret = lib_manager_allocate_module_instance(module_id, IPC4_INST_ID(ipc_config->id),
```

Comment:
Can you get the instance ID in an IPC version generic manner ?

Reply:
If it is going to support legacy modules it can't. Be assume that modules will be loadable without recompilation.

Reply:
sorry - I mean this file needs to be IPC version generic. We could pass this ID in from IPC specific client code or could have a generic macro (that does something different depending on IPC version Kconfig) that gets the ID.